gh-87389: Fix an open redirection vulnerability in http.server.

gpshead opened the pull request:
Fix an open redirection vulnerability in the `http.server` module when
an URI path starts with `//`.  Vulnerability discovered, and initial fix
proposed, by Hamza Avvan (@hamzaavvan).

Test authored and secondary mitigation by Gregory P. Smith [Google].

This PR takes over and replaces https://github.com/python/cpython/pull/24848.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->

gpshead committed:
gh-87389: Fix an open redirection vulnerability in http.server.

Fix an open redirection vulnerability in the `http.server` module when
an URI path starts with `//`.  Vulnerability discovered, and initial fix
proposed, by Hamza Avvan.

Test authored and secondary mitigation by Gregory P. Smith [Google].

gpshead committed:
A more defensive assert within the test.

gpshead committed:
Fix wording in some comments.

gpshead committed:
Add an explanatory comment in the test.

orsenthil commented on Lib/test/test_httpservers.py:
```diff
@@ -418,6 +418,28 @@ def test_undecodable_filename(self):
+        location = response.getheader('Location')
+        self.assertFalse(location.startswith('//'), msg=location)
+        self.assertEqual(location, f'/{attack_url.lstrip("/")}/',
+                msg='Expected Location header to start with a single / and '
```
expected_url = attack_url.lstrip("/") 

would have made it slightly more easy to follow IMO.

orsenthil reviewed (approved):
LGTM.

vstinner commented on Lib/test/test_httpservers.py:
```diff
@@ -418,6 +418,28 @@ def test_undecodable_filename(self):
+        os.mkdir(os.path.join(self.tempdir, 'existing_directory'))
+        # Canonicalizes to /tmp/tempdir_name/existing_directory which does
+        # exist and is a dir, triggering the 301 redirect and former bug.
+        attack_url = f'//python.org/..%2f..%2f..%2f..%2f..%2f../%0a%0d/../{self.tempdir_name}/existing_directory'
```
I suggest to build the whole expected location URL without lstrip(), for readability of the test itself:

```suggestion
        url = f'/python.org/..%2f..%2f..%2f..%2f..%2f../%0a%0d/../{self.tempdir_name}/existing_directory'
        attack_url = f'/{url}'  # //python.org/... (double slash prefix, no trailing slash)
        expected_location = f'{url}/'  # /python.org/.../ (single slash prefix, with trailing slash)
```

And replace the two asserts with just:

```
self.assertEqual(location, expected_location)
```

If you want to elaborate, you can add a comment, I don't think that it's needed to elaborate in the 3rd assertEqual() msg parameter.

vstinner commented on Lib/test/test_httpservers.py:
```diff
@@ -418,6 +418,28 @@ def test_undecodable_filename(self):
+        self.assertFalse(location.startswith('//'), msg=location)
+        self.assertEqual(location, f'/{attack_url.lstrip("/")}/',
+                msg='Expected Location header to start with a single / and '
+                'end with a / as this is a directory redirect.')
```
Would you mind to add a test with 3 slashes: ``///url`` ?

gpshead committed:
Address vstinner comments on the test.

make the base urls, attack urls, and expected_location more clear in the
test.  Adds an additional test for a triple-slash path to ensure we're
not only treating double slashes as special.

vstinner reviewed (approved):
LGTM.

While it's good to fix security issues, I would like to repeat the http.server warning in its documentation ;-)

> Warning: http.server is not recommended for production. **It only implements basic security checks.**

gpshead committed:
Add a test for absolute Request-URI and fixup to allow it.

Remove the send_head scheme & netloc neutering that I had added.
Unnecessary in the primary issue's case and actually causes a problem
in this other potential case.  It's unclear how often any clients
ever send an absolute-URI to a http server instead of a relative-URI
but as it works today we shouldn't break that.

gpshead committed:
Further improve test cases.

Adds a test for the non-attack as a starting all-ok check.
Makes the absolute scheme://netloc request-uri case more lenient
as all we care about there is that the scheme://netloc makes it
into the Location: header in that scenario.
